### PR TITLE
fix: DistributedEnforcer addPolicySelf and InternalEnforcer to use dispatcher in methods

### PR DESCRIPTION
**Pull request by alegueleres**
Hello, I am Alexandre, I currently work at [@ciandt](https://github.com/ciandt) as a Senior Software Engineer.

We are using jcasbin in my current project, we are having an issue updating multiple microservices instances.

Analyzing casbin I found Dispatcher and DistributedEnforcer that can handle our problem, but when I did some tests, I found these 2 issues:

1. DistributedEnforcer was ignoring a new rule that does not exist.
2. InternalEnforcer was not calling dispatcher on methods besides update policy method.

So this PR is to solve both problems and also I added unit test scenarios to cover these new lines of code.

Thank you.

edit: Fixed some codebeat issues reported

### CODE REVIEW

**Comment by hsluoyz** [comment]
@tangyang9464 @nodece @shy1st plz review

**Comment by alegueleres** [comment]
Hi, should I fix all the codebeat issues on this PR?

**Review by tangyang9464 — APPROVED** [review]
lgtm

File: src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java [hunk 1]
```diff
@@ -33,6 +35,11 @@ class InternalEnforcer extends CoreEnforcer {
      * addPolicy adds a rule to the current policy.
      */
     boolean addPolicy(String sec, String ptype, List<String> rule) {
+        if (mustUseDispatcher()) {
+            dispatcher.addPolicies(sec, ptype, singletonList(rule));
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java:39]:
```suggestion
            return dispatcher.addPolicies(sec, ptype, singletonList(rule));
```

File: src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java [hunk 2]
```diff
@@ -709,4 +709,8 @@ public boolean isAutoNotifyDispatcher() {
     public void setAutoNotifyDispatcher(boolean autoNotifyDispatcher) {
         this.autoNotifyDispatcher = autoNotifyDispatcher;
     }
+
+    public boolean mustUseDispatcher() {
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java:713]:
```suggestion
    protected boolean mustUseDispatcher() {
```

File: src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java [hunk 3]
```diff
@@ -71,6 +74,11 @@ boolean addPolicy(String sec, String ptype, List<String> rule) {
      * addPolicies adds rules to the current policy.
      */
     boolean addPolicies(String sec, String ptype, List<List<String>> rules) {
+        if (mustUseDispatcher()) {
+            dispatcher.addPolicies(sec, ptype, rules);
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java:78]:
```suggestion
            return dispatcher.addPolicies(sec, ptype, rules);
```

File: src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java [hunk 4]
```diff
@@ -115,6 +121,11 @@ public void buildIncrementalRoleLinks(Model.PolicyOperations op, String ptype, L
      * removePolicy removes a rule from the current policy.
      */
     boolean removePolicy(String sec, String ptype, List<String> rule) {
+        if (mustUseDispatcher()) {
+            dispatcher.removePolicies(sec, ptype, singletonList(rule));
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java:125]:
```suggestion
            return dispatcher.removePolicies(sec, ptype, singletonList(rule));
```

File: src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java [hunk 5]
```diff
@@ -225,6 +232,11 @@ boolean updatePolicy(String sec, String ptype, List<String> oldRule, List<String
      * removePolicies removes rules from the current policy.
      */
     boolean removePolicies(String sec, String ptype, List<List<String>> rules) {
+        if (mustUseDispatcher()) {
+            dispatcher.removePolicies(sec, ptype, rules);
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java:236]:
```suggestion
            return dispatcher.removePolicies(sec, ptype, rules);
```

File: src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java [hunk 6]
```diff
@@ -264,8 +274,13 @@ boolean removePolicies(String sec, String ptype, List<List<String>> rules) {
      * removeFilteredPolicy removes rules based on field filters from the current policy.
      */
     boolean removeFilteredPolicy(String sec, String ptype, int fieldIndex, String... fieldValues) {
+        if (mustUseDispatcher()) {
+            dispatcher.removeFilteredPolicy(sec, ptype, fieldIndex, fieldValues);
```
Comment (nodece) [inline src/main/java/org/casbin/jcasbin/main/InternalEnforcer.java:278]:
```suggestion
            return dispatcher.removeFilteredPolicy(sec, ptype, fieldIndex, fieldValues);
```

File: src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java (hunk 2 above)
Comment (alegueleres) [reply src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java]: Hi @nodece, great suggestion, Thank you.

Fixed.

**Pull request merged**